### PR TITLE
JAMES-2177 Runtime exceptions in IMAP and JMAP don't have MDC context

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/container/util-java8/src/test/java/org/apache/james/util/FluentFutureStreamTest.java [hunk 1]
```diff
@@ -33,6 +36,23 @@
 public class FluentFutureStreamTest {
 
     @Test
+    public void test() throws IOException {
```
Comment (aduprat) [inline server/container/util-java8/src/test/java/org/apache/james/util/FluentFutureStreamTest.java]: test ?

Comment (chibenwa) [reply server/container/util-java8/src/test/java/org/apache/james/util/FluentFutureStreamTest.java]: Oups. This was a debug statement to confirm execution order of cath close, and thus apply correctly MDC. Will remove.

File: metrics/metrics-api/src/main/java/org/apache/james/metrics/api/NoopMetricFactory.java [hunk 2]
```diff
@@ -55,4 +57,9 @@ public long stopAndPublish() {
             return 0;
         }
     }
+
+    @Override
+    public <T> T withMetric(String name, Supplier<T> operation) {
+        return operation.get();
```
Comment (rouazana) [inline metrics/metrics-api/src/main/java/org/apache/james/metrics/api/NoopMetricFactory.java]: is it still a Noop if it runs the operation?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 3]
```diff
@@ -83,25 +80,23 @@
...
-            timeMetric.stopAndPublish();
-        }
+        return metricFactory.withMetric(JMAP_PREFIX + METHOD_NAME.getName(),
+            () -> MDCBuilder.withMdc(
+                MDCBuilder.create()
+                    .addContext(MDCBuilder.ACTION, "GET_MAILBOXES")
+                    .addContext("accountId", mailboxesRequest.getAccountId())
+                    .addContext("mailboxIds", mailboxesRequest.getIds())
+                    .addContext("properties", mailboxesRequest.getProperties()),
+                () -> process(clientId, mailboxSession, mailboxesRequest)));
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: closing a closable should be the responsability of the caller, no? Now it's not obvious that it is well closed.

File: metrics/metrics-api/src/main/java/org/apache/james/metrics/api/NoopMetricFactory.java (hunk 2 above)
Comment (chibenwa) [reply metrics/metrics-api/src/main/java/org/apache/james/metrics/api/NoopMetricFactory.java]: imo it stays "transparent" to the operation then it makes sens to keep the NOOP naming.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java (hunk 3 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: That's why it's only the MDCBuilder being passed ;-)

MDCBuilder is not closeable, thus managing the closeable is no responsibility of the caller code, but of the callee (to rephrase me, the MDCBuilder::withMDC method will generate the Closeable and close it, thus there is no responsibility mismatch).

**Comment by strokyl** [comment]
I really like you PR, I push a commit : https://github.com/linagora/james-project/pull/1024/commits/62dd78e528842e5db0d429793f44f0fddcaa48af that is a suggestion to minimize the use of lambda and IMO improve the readability, if you don't like it feel free to remove it.
Also I am not proud of my naming ``wrapArround`` if you find better naming feel free to change it

**Comment by chibenwa** [comment]
For me that's :+1:

**Comment by chibenwa** [comment]
Merged

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**